src/Microsoft.ApplicationInsights.AspNetCore/RequestTrackingTelemetryModule.cs

```diff
@@ -0,0 +1,117 @@
+﻿namespace Microsoft.ApplicationInsights.AspNetCore
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Threading;
+    using Microsoft.ApplicationInsights.AspNetCore.DiagnosticListeners;
+    using Microsoft.ApplicationInsights.Extensibility;
+
+    /// <summary>
+    /// Telemetry module tracking requests using Diagnostic Listeners.
+    /// </summary>
+    public class RequestTrackingTelemetryModule : ITelemetryModule, IObserver<DiagnosticListener>, IDisposable
+    {
+        private TelemetryClient telemetryClient;
+        private IApplicationIdProvider applicationIdProvider;
+        private ConcurrentBag<IDisposable> subscriptions;
+        private List<IApplicationInsightDiagnosticListener> diagnosticListeners;
+        private bool isInitialized = false;
+        private readonly object lockObject = new object();
+        private TelemetryConfiguration configuration;
+
+
+        public RequestTrackingTelemetryModule() : this(null)
+        {            
+        }
+
+        public RequestTrackingTelemetryModule(IApplicationIdProvider applicationIdProvider)
+        {
+            this.applicationIdProvider = applicationIdProvider;
+            this.subscriptions = new ConcurrentBag<IDisposable>();
+            this.diagnosticListeners = new List<IApplicationInsightDiagnosticListener>();
+        }
+
+        /// <summary>
+        /// Initializes the telemetry module.
+        /// </summary>
+        /// <param name="configuration">Telemetry configuration to use for initialization.</param>
+        public void Initialize(TelemetryConfiguration configuration)
+        {
+            if (!this.isInitialized)
+            {
+                lock (this.lockObject)
+                {
+                    if (!this.isInitialized)
+                    {
+                        this.telemetryClient = new TelemetryClient(configuration);
+
+                        this.diagnosticListeners.Add
+                            (new HostingDiagnosticListener(this.telemetryClient, applicationIdProvider));
+
+                        this.diagnosticListeners.Add
+                            (new MvcDiagnosticsListener());
+
+                        this.subscriptions.Add(DiagnosticListener.AllListeners.Subscribe(this));
```

Comment:
check for null? Hypothetically Initialize may be called after/during Dispose

Reply:
Thanks. WIll do